src/wallet/coinselection.cpp

```diff
@@ -0,0 +1,349 @@
+#include "../tweak.h"
+#include "random.h"
+#include "wallet.h"
+
+// needed for tx validation
+#include "../main.h"
+#include "../script/sign.h"
+#include "../txmempool.h"
+const int MAX_SOLUTIONS = 2000; // The approximate maximum number of solutions we will find before giving up.
+// What's the maximum number of TXOs we should use as inputs to a transaction (if we have a choice).
+const int MAX_ELECTIVE_TXOS = 5;
+const int MAX_TXOS = 100; // What's the maximum number of TXOs to put in a transaction
+// how many satoshi's the coin selection is willing to be off by per iteration through the loop.  Makes it progressively
+// easier to find a solution
+// (now its a simple exponential) const long int LOOP_COST = 500;
+const int P2PKH_INPUT_SIZE = 72; // <Sig> <PubKey> OP_DUP OP_HASH160 <PubkeyHash> OP_EQUALVERIFY OP_CHECKSIG
+
+extern CTweak<unsigned int> maxCoinSelSearchTime;
+extern CTweak<unsigned int> preferredNumUTXO;
+
+// This holds a sorted map of coin amounts and the set of Txos required to get to that amount
+typedef std::multimap<CAmount, TxoGroup> TxoGroupMap;
+
+// How close do we want to get to targetValue (at first).  The longer it takes the more we add to this value.
+CAmount reasonableExcess(CAmount targetValue) { return targetValue / 1024; }
+// Create a group of Txos from a Txo index and another group.
+TxoGroup makeGroup(SpendableTxos::iterator i, const TxoGroup *prev = nullptr)
+{
+    TxoGroup ret;
+    if (prev != nullptr)
+        ret = *prev;
+    else
+        ret.first = 0;
+    // Don't add an element that already exists  -- allow this, bad solutions eliminated later
+    // assert(ret.second.find(i) == ret.second.end());
+    ret.first += i->first; // update the amount
```

Comment:
did you intend to remove this assert?

Reply:
yes.  This is again the performance tradeoff between allowing rare invalid solutions and removing them later and weeding them out in advance.  Note the comment says "-- allow this, bad solutions eliminated later".